pkg/apis/core/validation/validation.go

```diff
@@ -1020,7 +1019,11 @@ func validateDownwardAPIVolumeFile(file *core.DownwardAPIVolumeFile, fldPath *fi
 			allErrs = append(allErrs, field.Invalid(fldPath, "resource", "fieldRef and resourceFieldRef can not be specified simultaneously"))
 		}
 	} else if file.ResourceFieldRef != nil {
-		allErrs = append(allErrs, validateContainerResourceFieldSelector(file.ResourceFieldRef, &validContainerResourceFieldPathExpressions, fldPath.Child("resourceFieldRef"), true)...)
+		localValidContainerResourceFieldPathPrefixes := sets.NewString()
+		if utilfeature.DefaultFeatureGate.Enabled(features.DownwardAPIHugePages) {
```

Comment:
A couple options:
1. Rather than checking feature gate enablement inside this method, I'd recommend threading an options struct with a field that indicates whether hugepages downward API is allowed. At the top-level, when validating for Create API calls, populate that field based only on feature enablement. When validating for Update API calls, populate that field based on feature enablement OR presence of hugepages downward API data in the existing object.
2. Make this validation unconditional, have a dedicated conditional validation method for pod templates that distinguishes between create and update that adds forbidden errors if the feature is disabled and there is no existing object containing downward API hugepages data, and call that dedicated method from the create/update strategies of all the types that embed pod template. It looks like we had this until SCTP went to GA and it just got removed everywhere in https://github.com/kubernetes/kubernetes/commit/3a110546ed8ab319a3997b788889c943cdb6f05d#diff-3b1b49a75c00d94b4d54c16850cf4305863b36ccff086761e16e1ed920c8a02d :-(

Reply:
@liggitt i had to refresh my memory, but since pod updates do not allow this section of the pod spec to mutate, i don't think the above concerns apply and validation on create is sufficient.

Reply:
it's updates to the workload objects that would be an issue (and why the plumbing in https://github.com/kubernetes/kubernetes/commit/3a110546ed8ab319a3997b788889c943cdb6f05d#diff-3b1b49a75c00d94b4d54c16850cf4305863b36ccff086761e16e1ed920c8a02d touched so many things)

Reply:
ah, understood after further discussion.  option 1 seems simpler.

Reply:
after plumbing, i now realize top-level in this context meant the strategy object for anything that embeds a pod spec not top-level in validation logic!